### DISCUSSION TITLE
Quarkus 3.13.0: Maven builds use up to 4 times more resources than with 3.12.3

### DISCUSSION THREAD

**Discussion by admanzoni** [quarkusio/quarkus#42298] [Q&A] [answered]
Hi there!

I have just upgraded to from 3.12.3 to 3.13.0 and realized than the Maven build use up to 4 times more memory than before.
Has anyone noticed the same behaviour ?

[EDIT]
I did follow the recommendation and used `WithTestResource` instead of `QuarkusTestResource`.
By using `QuarkusTestResource` (deprecated way), my build use around 1/1.5 GB of memory
By using `WithTestResource`, my build use 6GB

[a comment by an automated account is not shown]

**Reply by aloubyansky** [reply]
@edeandrea fyi

**Reply by edeandrea** [reply, marked as answer]
All the new annotation does is change the default of `restrictToAnnotatedClass` from `false` to `true`. 

I think this means that the test resource, by default, is bound only to the single test class rather than all test classes. 

I'd be curious if you see the same thing if you use `@QuarkusTestResource(restrictToAnnotatedClass=true)`

My guess is that the behavior would be the same. Nothing else changed really

**Reply by gsmet** [reply]
Do we properly clean up the resources after the class is handled?

**Reply by edeandrea** [reply]
> Do we properly clean up the resources after the class is handled?

I honestly don't know. I didn't write any of the code that handles this. All I did was add the new annotation into the workflow that already existed.

**Reply by gsmet** [reply]
Well, sure, but changing the default means we need to be extra careful.

**Reply by edeandrea** [reply]
The way it was before was super confusing. Test resources (and any config a test resource may have injected) is available for all test classes, even if the test class wasn't annotated. The change itself was spawned by https://x.com/cowtowncoder/status/1736913697426338262

I'd be curious if the same observation is noted when using `@QuarkusTestResource(restrictToAnnotatedClass=true)`. You can always use `@WithTestResource(restrictAnnotatedClass=false)` to revert to previous behavior.

**Reply by gsmet** [reply]
I'm not sure we are talking of the same thing: I don't discuss the fact that it's a better default. I discuss the fact that if it's the new default, we need to make sure it's rock solid.

Now I have no idea what's going on here but it's definitely worth an investigation from my point of view.

**Reply by edeandrea** [reply]
As I mentioned before, I'd be curious if the same observation is noted when using `@QuarkusTestResource(restrictToAnnotatedClass=true)`.

@admanzoni ^^^

**Reply by admanzoni** [reply]
`restrictToAnnotatedClass` did the trick indeed :) Thanks !

I didn't pay attention to the change of the default value, my bad..

Thanks to all

**Reply by admanzoni** [reply]
@edeandrea Yes indeed `@QuarkusTestResource(restrictToAnnotatedClass=true)` also consumes more resources.

Btw, you guys are doing a great job, keep going !

**Comment by maxandersen** [comment]
Do you have a reproducer so we can be sure what issue/difference you are seeing?

**Reply by admanzoni** [reply]
Setting `restrictToAnnotatedClass ` to false did the trick

**Reply by maxandersen** [reply]
So just to be clear - we didn't change the default - a new annotation was introduced. 

You upgraded from 3.12 to 3.13 AND moved to the new @withtestresource, correct ?

**Reply by maxandersen** [reply]
Ah you updated the description. 

But yeah @edeandrea and @gsmet a default that costs 4x memory doesn't sound good. 

Is this expected ? Fixable?

**Reply by maxandersen** [reply]
How many test classes/methods do you have ? Curious if few, many, alot, ... :)

**Reply by admanzoni** [reply]
In that module, we have 17 test classes and between 6-7 methods each. 
The test resource is a Keycloak mock server.

**Reply by edeandrea** [reply]
@maxandersen https://github.com/quarkusio/quarkus/issues/37853 is where this was discussed/implemented. `@QuarkusTestResource` and the new `@WithTestResource` do the same thing. The only difference is the default value of `restrictToAnnotatedClass` flips from `false` with `@QuarkusTestResource` and is now `true` with `@WithTestResource`.

Because `@QuarkusTestResource(restrictToAnnotatedClass=true)` shows this difference, then something with the way `restrictToAnnotatedClass=true` works is most likely not right. I have no idea how that works. That pre-existed the new `@WithTestResource` annotation.

**Comment by gsmet** [comment]
@admanzoni would you have the ability to test https://github.com/quarkusio/quarkus/pull/42304 ?

**Reply by gsmet** [reply]
@admanzoni but first, could you downgrade your application to 3.12.3 and just update your `@QuarkusTestResource` with `restrictAnnotatedClass=true` and see how it behaves?

**Reply by admanzoni** [reply]
Sure, by downgrading to 3.12.3 and with `@QuarkusTestResource(value = KeycloakServerMock.class, restrictToAnnotatedClass = true)`, I also observe a increase of the memory consumption (about the same ratio x4)

**Reply by gsmet** [reply]
OK, thanks. So it's not my work who caused the issue, thanks for checking. I'll have a deeper look next week.

**Reply by mschorsch** [reply]
Same issue: https://github.com/quarkusio/quarkus/issues/42303#issuecomment-2266688411

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/issues/42303 has been fixed - it was a Keycloak DevService specific issue

**Comment by edeandrea** [comment]
Looks like another issue related? #42365

**Reply by edeandrea** [reply]
I ran the same test - downgraded to Quarkus 3.12.3 and switched to `@QuarkusTestResource` an tried with both `true` and `false` for `restrictToAnnotatedClass`.

https://github.com/quarkusio/quarkus/issues/42365#issuecomment-2273745570
